scripts/generate_motor_json.py

```diff
@@ -0,0 +1,472 @@
+*   -- Yuri Rage - 2024
+*
+* Abbreviated implementation of AP_MotorsMatrix and AP_MotorsMatrix_test:
+*   A bit of a brute force approach to parsing cpp, but it works (for now).
+*   JSON output should be functionally identical to AP_MotorsMatrix_test for
+*   the purposes of generating copter motor diagrams.
+*
+* Facilitates proper versioning of Copter motor diagrams completely
+*   within ardupilot_wiki (no need for added github tasks) by fetching cpp
+*   from Ardupilot/ardupilot and parsing it.
+*
+* May be somewhat fragile, so prefer running AP_MotorsMatrix_test
+*   in an Ardupilot/ardupilot environment to update JSON data.
+*
+****************************************************************************"""
+import argparse
+import json
+import math
+import pathlib
+import re
+from sys import stderr
+
+import requests
+from packaging.version import Version
+
+THIS_SCRIPT = pathlib.Path(__file__).resolve()
+DEFAULT_OUTPUT_FILE = THIS_SCRIPT.parent / "motor_diagram_data/AP_Motors_test.json"
+AP_GITHUB_OWNER = "ArduPilot"
+AP_GITHUB_REPO = "ardupilot"
+AP_COPTER_BRANCH = "Copter-"
+AP_MOTORS_CLASS_H = "libraries/AP_Motors/AP_Motors_Class.h"
+AP_MOTORS_MATRIX_CPP = "libraries/AP_Motors/AP_MotorsMatrix.cpp"
+# skip frame types that are not supported by AP_MotorsMatrix_test
+SKIP_FRAME_TYPES = ["MOTOR_FRAME_TYPE_Y4"]
+
+
+def fetch_branches(owner, repo, per_page=100):
+    """
+    fetch branch data (json) using github api
+    """
+    branches = []
+    url = f"https://api.github.com/repos/{owner}/{repo}/branches"
+    params = {"per_page": per_page}
+    page = 1
+    while url:
+        stderr.write(f"\rFetching branches for '{owner}/{repo}' [page {page: >2}] ... ")
+        stderr.flush()
+        try:
+            response = requests.get(url, params=params)
+            response.raise_for_status()
+            data = response.json()
+            branches.extend(
+                [
+                    {"name": branch["name"], "sha": branch["commit"]["sha"]}
+                    for branch in data
+                ]
+            )
+        except requests.RequestException:
+            print("request error.", file=stderr)
+            return branches
+
+        if "next" in response.links:
+            url = response.links["next"]["url"]
+            page += 1
+        else:
+            break
+
+    print("done.", file=stderr)
+    return branches
+
+
+def get_latest_branch(owner, repo, prefix=""):
+    """
+    fetch branch data and return the latest branch version matching the prefix
+    returns a dict with 'name' and 'sha' keys; None if no matches found
+    """
+    re_prefix = f"^{prefix}"
+    if prefix is None or prefix == "":
+        re_prefix = ""
+    version_regex = re.compile(rf"{re_prefix}(\d+\.\d+\.\d+|\d+\.\d+)")
+    branches = fetch_branches(owner, repo)
+    version_branches = [
+        (branch, Version(match.group(1)))
+        for branch in branches
+        if (match := version_regex.search(branch["name"]))
```

Comment:
I think this is what is upsetting CI (`:=`)

Reply:
Indeed. Apparently we are a little behind on our Python operator linting. Fixed. And thank you!